### DISCUSSION TITLE
Trying to serve manipulated image.

### DISCUSSION THREAD

**Discussion by shivangrathore** [rwf2/Rocket#1722] [Questions] [answered]
minimal reproducible code.

```rs
#[rocket::async_trait]
impl<'r> Responder<'r, 'static> for Image {
    fn respond_to(self, _: &'r Request<'_>) -> response::Result<'static> {
        let mut response = Response::new();
        response.set_sized_body(self.0.len(), io::Cursor::new(self.0));
        response.set_header(ContentType::from_extension("png").unwrap());
        return Ok(response);
    }
}

#[get("/do")]
async fn do_image() -> Image{
    let img = image::open("static/image.png").unwrap();
    let bytes = img.to_bytes();
    println!("{:?}", bytes);
    return Image(img.to_bytes().to_vec());
}
```

I am not soo good at rust, But above code returns a blank image. Any Ideas why?

**Comment by SergioBenitez** [comment]
It's hard to say without any more information. Perhaps showing us some logs would help?

Here are some general tips:



If you really just want to return a file, use `FileServer`. No need to write any logic.

If for whatever reason you don't want to use that, your route should look like:

```rust
#[get("/do")]
async fn image() -> Option<NamedFile> {
    NamedFile::open("static/images.png").await.ok()
}
```

You may also want to use `rocket::fs::relative!()`.

If you really need to read the bytes from the image into a buffer, then your `Image` and route should probably look like:

```rust
#[derive(Responder)]
#[response(content_type = "image/png")]
struct Image(Vec<u8>);

#[get("/do")]
async fn image() -> Result<Image> {
    let img = image::open("static/image.png")?;
    Ok(Image(img..to_vec()))
}
```

**Comment by jebrosen** [comment, marked as answer]
In addition to @SergioBenitez 's answer explaining a better option, here is why what you already tried doesn't work:

```rust
    let img = image::open("static/image.png").unwrap();
    let bytes = img.to_bytes();
```

[`to_bytes`](https://docs.rs/image/0.23.14/image/enum.DynamicImage.html#method.to_bytes) returns "a copy of this image’s pixels as a byte vector." Generally, browsers don't support raw image pixels; it must first be encoded to an image format such as JPEG or PNG.

Here is a similar question, which also used `image::open`: https://github.com/SergioBenitez/Rocket/issues/1496#issuecomment-752050698

**Reply by shivangrathore** [reply]
> In addition to @SergioBenitez 's answer explaining a better option, here is why what you already tried doesn't work:
> 
> ```rust
>     let img = image::open("static/image.png").unwrap();
>     let bytes = img.to_bytes();
> ```
> 
> [`to_bytes`](https://docs.rs/image/0.23.14/image/enum.DynamicImage.html#method.to_bytes) returns "a copy of this image’s pixels as a byte vector." Generally, browsers don't support raw image pixels; it must first be encoded to an image format such as JPEG or PNG.
> 
> Here is a similar question, which also used `image::open`: [#1496 (comment)](https://github.com/SergioBenitez/Rocket/issues/1496#issuecomment-752050698)

Thanks, the mentioned comment worked for me.
